EIPS/eip-7160.md

```diff
@@ -0,0 +1,165 @@
+---
+eip: 7160
+title: ERC-721 Multi-Metadata Extension
+description: "Multiple metadata URIs per token, with the option to pin a primary URI."
+author: 0xG (@0xGh), Marco Peyfuss (@mpeyfuss)
+discussions-to: https://ethereum-magicians.org/t/erc721-multi-metadata-extension/14629
+status: Draft
+type: Standards Track
+category: ERC
+created: 2023-06-09
+requires: 165, 721
+---
+
+## Abstract
+
+This EIP proposes an extension to the [ERC-721](./eip-721.md) standard to support multiple metadata URIs per token. It introduces a new interface, `IERC721MultiMetadata`, which provides methods for accessing the metadata URIs associated with a token, including a pinned URI index and a list of all metadata URIs. The extension is designed to be backward compatible with existing `ERC721Metadata` implementations.
+
+## Motivation
+
+The current [ERC-721](./eip-721.md) standard allows for a single metadata URI per token. However, there are use cases where multiple metadata URIs are desirable, such as when a token represents a collection of (cycling) assets with individual metadata, historic token metadata, collaborative and multi-artist tokens, evolving tokens. This extension enables such use cases by introducing the concept of multi-metadata support.
+
+The primary reason for having a multi-metadata standard versus using the existing `tokenURI` method with some token-specific functions to change the active URI is that dapps and marketplaces don't have a mechanism to infer and display all the token URIs. Furthermore the non-pinned URIs might be equally important: imagine a colaborative token with multiple artworks per token. Multi-metadata tokens can be displayed in a carousel, gallery, list, etc.
+
+## Specification
+
+The `IERC721MultiMetadata` interface:
+
+```solidity
+interface IERC721MultiMetadata /* is IERC721 */ {
+  event TokenUriPinned(uint256 indexed tokenId, uint256 indexed index, address indexed sender);
+  event TokenUriUnpinned(uint256 indexed tokenId, address indexed sender);
+
+  function tokenURIs(uint256 tokenId) external view returns (uint256 index, string[] memory uris);
```

Comment:
@SamWilsn initially my idea was to propose only the addition of this method as it would allow dapps to retrieve all the uris and the current index (whether it is pinned or not).

However I wonder if it is a good idea to provide a standard interface to dapps/marketplaces to implement pinning, unpinning and related events. Should they decide to adopt this EIP I assume that it would make sense for them to have a standard way to offer pinning/unpinning via UI. By the way we'll try to get their opinion too.

Reply:
@0xGh don't the functions on lines 34 and 35 accomplish this?

Reply:
Yes they do. Just sharing some context and thoughts based on what we discussed in private.